src/main/java/org/folio/service/invoice/PoLineInvoiceLineHolderBuilder.java

```diff
@@ -0,0 +1,108 @@
+import org.folio.rest.jaxrs.model.FundDistribution;
+import org.folio.service.finance.FiscalYearService;
+import org.folio.service.finance.transaction.EncumbranceService;
+
+import java.util.Map;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.HashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class PoLineInvoiceLineHolderBuilder {
+  private static final List<InvoiceLine.InvoiceLineStatus> EDITABLE_STATUSES = List.of(InvoiceLine.InvoiceLineStatus.OPEN, InvoiceLine.InvoiceLineStatus.REVIEWED);
+  private final FiscalYearService fiscalYearService;
+  private final InvoiceLineService invoiceLineService;
+  private final EncumbranceService encumbranceService;
+
+  public PoLineInvoiceLineHolderBuilder(FiscalYearService fiscalYearService, InvoiceLineService invoiceLineService,
+                                        EncumbranceService encumbranceService) {
+    this.fiscalYearService = fiscalYearService;
+    this.invoiceLineService = invoiceLineService;
+    this.encumbranceService = encumbranceService;
+  }
+
+  public CompletableFuture<PoLineInvoiceLineHolder> buildHolder(CompositePoLine compOrderLine, RequestContext requestContext) {
+    PoLineInvoiceLineHolder holder = new PoLineInvoiceLineHolder(compOrderLine);
+    return invoiceLineService.getInvoiceLinesByOrderLineId(compOrderLine.getId(), requestContext)
+      .thenCompose(invoiceLines -> CollectionUtils.isEmpty(invoiceLines) ? CompletableFuture.completedFuture(holder) :
+        CompletableFuture.completedFuture(getOpenOrReviewedInvoiceLines(holder.getInvoiceLines()))
+          .thenAccept(holder::withOpenOrReviewedInvoiceLines)
+          .thenCompose(aVoid -> validateAndRetrievePaidInvoiceLines(compOrderLine, holder.getInvoiceLines(), requestContext))
+          .thenAccept(holder::withCurrentYearPaidInvoiceLines)
+          .thenApply(aVoid -> holder));
+  }
+
+  private CompletableFuture<List<InvoiceLine>> validateAndRetrievePaidInvoiceLines(CompositePoLine compOrderLine, List<InvoiceLine> invoiceLines, RequestContext requestContext) {
+    Optional<FundDistribution> optionalFundDistribution = compOrderLine.getFundDistribution().stream().findFirst();
+    if (optionalFundDistribution.isPresent()) {
+      String poLineFundId = optionalFundDistribution.get().getFundId();
+      return getCurrentFiscalYearInvoiceLines(poLineFundId, invoiceLines, requestContext)
+        .thenApply(currentYearInvoiceLines -> {
+          validateInvoiceLinesStatus(currentYearInvoiceLines);
+          return getPaidInvoiceLines(currentYearInvoiceLines);
+        });
+    } else {
+      return CompletableFuture.completedFuture(null); //Should ask Dennis what we should do in case of POL doesn't have fund distribution
+    }
+  }
+
+  private CompletableFuture<List<InvoiceLine>> getCurrentFiscalYearInvoiceLines(String poLineFundId, List<InvoiceLine> invoiceLines, RequestContext requestContext) {
```

Comment:
please add comments to sections/blocks of this method. Comments will describe what each block is doing in english.

Reply:
Added. I hope this will be enough. Thank you